### PR TITLE
Reworked the code

### PR DESCRIPTION
**Pull request by u-fm**
Basically based on established Japanese rules.

- Added 2/3 player mode (no president/beggar, only minister/peasant)
- Fixed passing rules (passed players can play cards later when their turn comes again)
- Added Dalmuti style scoring option (1 per player beaten) and deflated the score (5 -> 1)
- Jokers are now not almighty cards which beat everything, but the card above 2/3 (revolution) or a wild
- Allow 5 or more cards to cause Revolution
- Compressed variants as rule sets to reduce player pool split due to the number of options, removed card rank option for the same reason as it does not change the game play at all
- Remove limited deal option (this makes the game highly dependent on luck and severely increases taxer advantage)
- Added Sequence option (same suit of consecutive cards)
- Added Eight end option (playing an 8 ends the trick immediately)
- Added Jack back option (playing a Jack triggers a temporary revolution during that trick)
- Added Suit lock option (playing the same suit combination as the previous player 'locks' the play during that trick to that suit combination only)
- Added Illegal finish option (finishing the hand by the highest card or the Joker makes the player the lowest rank)
- Added Downfall option (if the President fails to go out first, they become the lowest rank)
- Redesigned UI (displays number of cards/round rank more clearly and shows revolution status by reordering cards and displaying related signs)

### CODE REVIEW

**Comment by boekabart** [comment]
> Added 2/3 player mode (no president/beggar, only minister/peasant)

IMO should be an option to choose between passing 1 or 2 cards in this case.

**Comment by boekabart** [comment]
> Fixed passing rules (passed players can play cards later when their turn comes again)

Like! But should be an option I think.

File: gameinfos.inc.php [hunk 1]
```diff
@@ -32,108 +27,99 @@
...
+'players' => [2, 3, 4, 5, 6, 7, 8],
 
 // Suggest players to play with this number of players. Must be null if there is no such advice, or if there is only one possible player configuration.
+// NB: the automatic lobby will try first the lowest number of players if this is not specified. So you _have to_ specify this parameter if the lowest player number is not compatible with the default options.
 'suggest_player_number' => null,
 
 // Discourage players to play with these numbers of players. Must be null if there is no such advice.
-'not_recommend_player_number' => null,
-// 'not_recommend_player_number' => array( 2, 3 ),      // <= example: this is not recommended to play this game with 2 or 3 players
+'not_recommend_player_number' => [2, 3],
```
Comment (boekabart) [inline gameinfos.inc.php:37]: and 5,6,7,8 in my experience. Only 4 is great :)

**Comment by u-fm** [comment]
> > Added 2/3 player mode (no president/beggar, only minister/peasant)
> 
> IMO should be an option to choose between passing 1 or 2 cards in this case.

As passing 2 makes the game too lopsided, common 2/3p variants adopt passing 1 card only.


> > Fixed passing rules (passed players can play cards later when their turn comes again)
> 
> Like! But should be an option I think.

I considered it, but in fact many players think this is a bug, not a variant difference.


> and 5,6,7,8 in my experience. Only 4 is great :)

Setting 4 as the suggest_player_number would be better I guess.

**Comment by quaresma95** [comment]
> I considered it, but in fact many players think this is a bug, not a variant difference

I agree i hink this should be an option too, this could be considered by other usées.

> Jokers are now not almighty cards which beat everything, but the card above 2/3 (revolution) or a wild

This too, i think more the game is customisable better it is.
Like everyone can play like they want.

> Remove limited deal option (this makes the game highly dependent on luck and severely increases taxer advantage)

What do you mean ? 
By the way, thanks for you contribution and your ideas. 🙂

**Comment by u-fm** [comment]
In fact, adding tons of minor variants is not a good idea in BGA as it confuses players, splits player pool into more pieces and hinders matchmaking.
To play a game on BGA players must find other players, so 'more customizable' does not always mean 'better'. 
That's why I made 'Rule set' option as a broader category.
Check BGA Hand and Foot's option list, then you'll know what I'm saying.

Also, what I removed in this version are all uncommon or unpopular variants compared to others like Revolution or Skip.
Some players even treat them as bugs. For example:
https://boardgamearena.com/bug?id=25073
https://boardgamearena.com/bug?id=57488

**Comment by quaresma95** [comment]
> To play a game on BGA players must find other players, so 'more customizable' does not always mean 'better'.

Ok I see what you mean, i agree small variants should not be an option. Like the revolution with 5 cards.

> Also, what I removed in this version are all uncommon or unpopular variants compared to others like Revolution or Skip.
For them it's bug, for the players who don't complain it's normal. 
For me it's not the same game without this rule. This change the all game.

> Remove limited deal option (this makes the game highly dependent on luck and severely increases taxer advantage)
The 52 cards are random ? You can have seveen 4 ?

**Comment by u-fm** [comment]
> For them it's bug, for the players who don't complain it's normal.

I couldn't find any major credible rule sources citing those rules as common rules. Imo forcing them is not 'normal'.
For example:
https://www.britannica.com/topic/president-card-game (no permanent pass, no almighty joker, no limited deal)
https://bicyclecards.com/how-to-play/presidents (no permanent pass or limited deal. this one does not include jokers at all)
https://www.pagat.com/climbing/president.html (the default descriptions of pass/joker are in line with my version, no limited deal)
No need to mention Japanese rule sources, obviously.

Considering the player pool size, getting more than 5 upvotes even though they are not 'bugs' per se means these are quite odd even without taking duplicate reports into account.
Making them as options is also questionable as I don't think making Jokers as almighty cards or keeping passed status until the trick end would make the game better, and as above, to prevent player pool splits I don't want to add unpopular variants.

> For me it's not the same game without this rule. This change the all game.

The current implementation forced uncommon rules and they have to be modified to meet the players' usual expectations.
For example, let's assume a President implementation uses multiple decks.
Changing it to a single deck surely changes the game play, but this doesn't mean it should not be modified.

> The 52 cards are random ? You can have seveen 4 ?

No.  The card composition has nothing to do.
The deal will be always the same as the previous default deal option (no limit on cards per player) and that's all.

**Comment by u-fm** [comment]
Closing the request as there have been no further replies

**Comment by ruben217** [comment]
Sorry i'm in vacation. 
Your PR will be merged when i will get back home.

**Pull request closed without merging**